libraries/net/node.cpp

```diff
@@ -4192,15 +4201,24 @@ namespace graphene { namespace net { namespace detail {
       initiate_connect_to(new_peer);
     }
 
-    peer_connection_ptr node_impl::get_connection_to_endpoint( const fc::ip::endpoint& remote_endpoint )
-    {
+   peer_connection_ptr node_impl::get_active_connection_to_endpoint( const fc::ip::endpoint& remote_endpoint)
+   {
       VERIFY_CORRECT_THREAD();
       for( const peer_connection_ptr& active_peer : _active_connections )
```

Comment:
There are several ways to go about this, and I am happy to do any one of them.

1. Stay with the way it is. A linear search through a vector is often faster than map for small to medium collections. So if you can put up with the ugly code, it will probably be faster.
1. Add a hash and equals to peer_connection_ptr, so that find() will work given an endpoint.
1. Use Boost.Multiindex to sort the collection, so that we can filter by more than just an endpoint (There are similar searches on node_id).

Or other ways that I haven't thought of. Let me know what you think is best.

Reply:
Here is more information that may help answer the questions above:

There are several collections within BitShares-Core’s node.cpp that have to do with connections to other nodes and their status. The most widely used is a collection of peers that we are actively connected to.

These collections (including _active_connections) are currently implemented as std::unordered_set<peer_connection_ptr>.

A peer_connection_ptr holds information about the peer. Things such as their endpoint (i.e. IP address and port), node id, and more.

A std::unordered_set is a collection that does not allow duplicates, and because it is unordered, has improved insert/delete speed as compared to an ordered set. Note that searches on this collection search for the hash of the object itself. So searches for something within the object are not possible, other than linear searches (Grab the object, look to see if it is the one you want. If not, move on.).

There are a few places in the code where it would be beneficial to do a search on information within the object. Searching by node id or endpoint are immediately necessary.

Unless we want to consider a linear search, unordered_set is not the collection that should be used here. So what should it be replaced with? To answer that question, we need to know how the collection will be used.

the _active_connections collection is often iterated through, added to, deleted from, and also searched by (1) the peer_connection_ptr itself, (2) the endpoint, and (3) the node_id.

If we want to support non-linear searching by different keys, the boost::MultiIndex would be an option. If we could eliminate the uses of all but one of the searches, we could use a “standard” container instead.

Searching by the peer_connection_ptr is used in many places. Searching by node_id always relates to firewall checking. Searching by endpoint is used to (1) verify we are not already connected to this endpoint, (2) check to see if the connection process is in progress, and (3) verify that an endpoint is in the “advertise” list and is active.

Eliminating searching by the peer_connection_ptr is not impossible. Since we have the object, we could search by a unique data element within. Eliminating searching by node_id means refactoring the firewall checking code. Eliminating searching by endpoint is more difficult, as we can only rely on node_id after the connection is established.

So the question becomes: Do we want to (a) move _active_connections to a boost:MultiIndex and allow for searching by node_id or endpoint (b) remove node_id searching and move _active_connections to something like a std::map or std::unordered_map , (c) do linear searching in the current std::unordered_set, or (d) something I have not thought of.

Reply:
I think we shouldn't change this in the current PR.

Reply:
@abitmore if you agree with pmconrad, perhaps an issue should be created?

Reply:
I agree.